### PR TITLE
Annotate nullability of reference types

### PR DESCRIPTION
**Pull request by atifaziz**
This PR [enables the nullable context](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references#nullable-contexts) for the entire project and annotates all reference types with respect to their nullability using [nullable references](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references) that are being introduced with C# 8.

There are a few cases marked with `TODO(nullable)` that I had trouble reasoning about.

Merging this PR would mean that the code base will be moved to C# 8. Is it too early to impose?

### CODE REVIEW

**Comment by atifaziz** [comment]
[Build is failing on AppVeyor](https://ci.appveyor.com/project/raboof/morelinq/builds/24280929) because Visual Studio 2019 image is not yet ready.

**Comment by jskeet** [comment]
I'm unlikely to be able to find time to do this any time soon, I'm afraid - I've got a lot on my plate right now.
However, you might want to grab code from the Noda Time appveyor script: https://github.com/nodatime/nodatime/blob/master/appveyor.yml

**Comment by atifaziz** [comment]
> I'm unlikely to be able to find time to do this any time soon, I'm afraid - I've got a lot on my plate right now.

Don't we all. :wink: I'm not in a big hurry to merge this if you think you'll find some time in a few days. I don't expect it to be a huge effort since all warnings are addressed and tests are passing. There's just a couple of corner cases but which might be just compiler bugs for now unless I've misunderstood something. It would have been great to get your quick glance especially with [your experience with the same for NodaTime](https://github.com/nodatime/nodatime/pull/1240). I'll remove you as a reviewer for now but if you'd like to be back on the panel then let me know.

> However, you might want to grab code from the Noda Time appveyor script: https://github.com/nodatime/nodatime/blob/master/appveyor.yml

Thanks! Using the **Visual Studio 2019 Preview** image with 52406abbd3c56013aa6d459a914583fe5408bac4 seems to have done the trick for MSBuild meanwhile though it still seems to be [failing due to SDK version mismatch](https://ci.appveyor.com/project/raboof/morelinq/builds/24282465).

File: .travis.yml [hunk 1]
```diff
@@ -5,24 +5,23 @@ os:
 osx_image: xcode8.3
 solution: MoreLinq.sln
 mono: 5.0.1
-dist: trusty
+dist: xenial
```
Comment (fsateler) [inline .travis.yml]: Any reason to prefer xenial instead of bionic?

File: MoreLinq/MoreLinq.csproj [hunk 2]
```diff
@@ -115,7 +115,8 @@
     <VersionPrefix>3.2.0</VersionPrefix>
     <Authors>MoreLINQ Developers.</Authors>
     <TargetFrameworks>net451;netstandard1.0;netstandard2.0</TargetFrameworks>
-    <LangVersion>7.3</LangVersion>
+    <LangVersion>8</LangVersion>
+    <NullableContextOptions>enable</NullableContextOptions>
```
Comment (fsateler) [inline MoreLinq/MoreLinq.csproj]: This should be `<Nullable>enabled</Nullable>` now

File: MoreLinq/Pad.cs [hunk 3]
```diff
@@ -111,11 +111,11 @@ public static IEnumerable<TSource> Pad<TSource>(this IEnumerable<TSource> source
             if (source == null) throw new ArgumentNullException(nameof(source));
             if (paddingSelector == null) throw new ArgumentNullException(nameof(paddingSelector));
             if (width < 0) throw new ArgumentException(null, nameof(width));
-            return PadImpl(source, width, default, paddingSelector);
+            return PadImpl(source, width, default!, paddingSelector);
```
Comment (fsateler) [inline MoreLinq/Pad.cs]: Why? Shouldn't PadImpl accept a `T?`

File: MoreLinq/Partition.cs [hunk 4]
```diff
@@ -174,11 +174,11 @@ public static (IEnumerable<T> True, IEnumerable<T> False)
         /// </returns>
 
         public static TResult Partition<TKey, TElement, TResult>(this IEnumerable<IGrouping<TKey, TElement>> source,
-            TKey key, IEqualityComparer<TKey> comparer,
+            TKey key, IEqualityComparer<TKey>? comparer,
             Func<IEnumerable<TElement>, IEnumerable<IGrouping<TKey, TElement>>, TResult> resultSelector)
         {
             if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
-            return PartitionImpl(source, 1, key, default, default, comparer,
+            return PartitionImpl(source, 1, key, default!, default!, comparer,
```
Comment (fsateler) [inline MoreLinq/Partition.cs]: Again, I don't understand why forbid reference `T`.

File: .travis.yml (hunk 1 above)
Comment (atifaziz) [reply .travis.yml:8]: None in particular except that xenial was the least required version.

File: MoreLinq/MoreLinq.csproj (hunk 2 above)
Comment (atifaziz) [reply MoreLinq/MoreLinq.csproj]: Fixed by @Orace in 9723435a0f0318d90ccec7085a15c24babcf0f4f.

**Comment by atifaziz** [comment]
I've integrated all the contributions by @Orace and @moh-hassan from PR #713 into here.

File: MoreLinq/Pad.cs (hunk 3 above)
Comment (atifaziz) [reply MoreLinq/Pad.cs:113]: You can't express `T?` for `T`. See following passage in [Generic definitions and nullability](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-migration-strategies#generic-definitions-and-nullability) section of the [nullable migration strategies](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-migration-strategies documentation):

> What it does mean is that you can't use `T?` in a generic class or method declaration without constraints. For example, [`Enumerable.FirstOrDefault<TSource>(IEnumerable<TSource>)`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.firstordefault#System_Linq_Enumerable_FirstOrDefault__1_System_Collections_Generic_IEnumerable___0__) won't be changed to return `T?`.

File: MoreLinq/Partition.cs (hunk 4 above)
Comment (atifaziz) [reply MoreLinq/Partition.cs:181]: Ditto [comment](https://github.com/morelinq/MoreLINQ/pull/582#discussion_r469376544).

**Comment by atifaziz** [comment]
Odd that [_non-Windows_ builds are failing](https://travis-ci.org/github/morelinq/MoreLINQ/builds/717307079) due to (nullability) [compilation errors](https://travis-ci.org/github/morelinq/MoreLINQ/jobs/717307080#L1604-L1625)… 🤔

```
PendNode.cs(66,27): error CS8653: A default expression introduces a null value when 'T' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
PendNode.cs(67,27): error CS8653: A default expression introduces a null value when 'T' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
PendNode.cs(68,27): error CS8653: A default expression introduces a null value when 'T' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
PendNode.cs(69,27): error CS8653: A default expression introduces a null value when 'T' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
CountBy.cs(97,43): error CS8653: A default expression introduces a null value when 'TKey' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
Flatten.cs(126,29): error CS8600: Converting null literal or possible null value to non-nullable type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
GroupAdjacent.cs(270,25): error CS8653: A default expression introduces a null value when 'TKey' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
IndexBy.cs(73,73): error CS8653: A default expression introduces a null value when 'TSource' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
PartialSort.cs(233,27): error CS8653: A default expression introduces a null value when 'TKey' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
ScanBy.cs(93,39): error CS8653: A default expression introduces a null value when 'TKey' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
ScanBy.cs(94,44): error CS8653: A default expression introduces a null value when 'TState' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
PendNode.cs(66,27): error CS8653: A default expression introduces a null value when 'T' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
PendNode.cs(67,27): error CS8653: A default expression introduces a null value when 'T' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
PendNode.cs(68,27): error CS8653: A default expression introduces a null value when 'T' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
PendNode.cs(69,27): error CS8653: A default expression introduces a null value when 'T' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
CountBy.cs(97,43): error CS8653: A default expression introduces a null value when 'TKey' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
Flatten.cs(126,29): error CS8600: Converting null literal or possible null value to non-nullable type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
GroupAdjacent.cs(270,25): error CS8653: A default expression introduces a null value when 'TKey' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
IndexBy.cs(73,73): error CS8653: A default expression introduces a null value when 'TSource' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
PartialSort.cs(233,27): error CS8653: A default expression introduces a null value when 'TKey' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
ScanBy.cs(93,39): error CS8653: A default expression introduces a null value when 'TKey' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
ScanBy.cs(94,44): error CS8653: A default expression introduces a null value when 'TState' is a non-nullable reference type. [/home/travis/build/morelinq/MoreLINQ/MoreLinq/MoreLinq.csproj]
```

**Comment by atifaziz** [comment]
> Odd that [_non-Windows_ builds are failing](https://travis-ci.org/github/morelinq/MoreLINQ/builds/717307079) due to (nullability) [compilation errors](https://travis-ci.org/github/morelinq/MoreLINQ/jobs/717307080#L1604-L1625)… 🤔

The Windows build was working because it was [effectively rolling forward to SDK 3.1.302](https://ci.appveyor.com/project/raboof/morelinq/builds/34622030#L258) whereas [other platform builds were using 3.1.102](https://travis-ci.org/github/morelinq/MoreLINQ/jobs/717307080#L1499) and the former no longer produces the warnings/errors that the latter does. Hopefully cca922759fccd930d978773b3f6bb814334321a7 will sort this out.

File: MoreLinq/Pad.cs (hunk 3 above)
Comment (moh-hassan) [reply MoreLinq/Pad.cs:114]: @atifaziz , You can add the ` [return:MaybeNull]` annotation to the method return using the  Nullable [Attributes](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis#specify-post-conditions-maybenull-and-notnull) to express T?.

Comment (atifaziz) [reply MoreLinq/Pad.cs:114]: @moh-hassan `MaybeNull` is a post-condition and we're not discussing the return type here.

**Comment by atifaziz** [comment]
@fsateler, @leandromoh I've requested your reviews, but I imagine you might not find time in the coming days. On the other hand, this PR makes sweeping changes that I'd like to merge soon to avoid divergence and merge conflicts; it's also been open for far too long. There should be no behavioural changes so I don't expect introduction of any bugs. That said, I did refactor some code to be more honest in the face of nullability. I will merge this PR by this weekend if I don't hear back on premise that there will always be time to make improvements. If you get round to reviewing the changes at a later time then I'd suggest just dropping your comments here (although that would unusual for a PR that will have been closed) or opening new new PRs to propose your changes/improvements. Needless to say, reviews are equally welcome from everyone else who has contributed or commented on this PR so far.

File: MoreLinq/CountBy.cs [hunk 5]
```diff
@@ -78,51 +78,35 @@ static partial class MoreEnumerable
...
-                    {
-                        if (key == null)
-                        {
-                            i = nullIndex.GetValueOrDefault();
-                            return nullIndex.HasValue;
-                        }
-
-                        return dic.TryGetValue(key, out i);
-                    }
+                    var dic = new Collections.Dictionary<TKey, int>(cmp);
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply MoreLinq/CountBy.cs:81]: It's actually related. This refactoring, as well as any other, was strictly done to avoid introducing uses of dammit operator (`!`) if it could be helped. I started by refactoring the local function into a normal static method, where the parameter list grew. Then I saw the same pattern and issue in `ScanBy` and decided to extract to a separate type.

File: MoreLinq/CountBy.cs [hunk 6]
```diff
@@ -94,17 +94,17 @@ bool TryGetIndex(TKey key, out int i)
...
                     var index = 0;
 
                     foreach (var item in source)
                     {
                         var key = keySelector(item);
 
                         if (// key same as the previous? then re-use the index
-                            havePrevKey && cmp.GetHashCode(prevKey) == cmp.GetHashCode(key)
-                                         && cmp.Equals(prevKey, key)
+                            prevKey is (true, {} pk)
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply MoreLinq/CountBy.cs:93]: I feel it belongs in this PR because pattern matching here avoids using dammit (`!`) and…

> This is a hot path and changing simple calls for pattern matching could reduce performance.

…the cost seems to be [negligible](https://gist.github.com/atifaziz/b1fcf40a266ae9fecff0b33a77c5e06b) ([see also on SharpLab](https://sharplab.io/#gist:b1fcf40a266ae9fecff0b33a77c5e06b)) as the following test:

```c#
using System;

public class C {
    public void M((bool, object) x) {
        if (x is (true, {} obj)) {
            Console.WriteLine(obj);
        }
    }
    public void N(bool f, object obj) {
        if (f) {
            Console.WriteLine(obj);
        }
    }
}
```

shows it adds just 6 bytes of x64 instructions (assuming that's a satisfactory measure):

```asm
; Core CLR v4.700.20.26901 on amd64

C..ctor()
    L0000: ret

C.M(System.ValueTuple`2<Boolean,System.Object>)
    L0000: cmp byte ptr [rdx+8], 0
    L0004: je short L001b
    L0006: mov rcx, [rdx]
    L0009: test rcx, rcx
    L000c: je short L001b
    L000e: mov rax, System.Console.WriteLine(System.Object)
    L0018: jmp rax
    L001b: ret

C.N(Boolean, System.Object)
    L0000: test dl, dl
    L0002: je short L0014
    L0004: mov rcx, r8
    L0007: mov rax, System.Console.WriteLine(System.Object)
    L0011: jmp rax
    L0014: ret
```

If you still feel strongly about this going into a separate PR then perhaps you open that as a new issue and we revert via a PR. What's your final verdict?

File: MoreLinq/Experimental/Await.cs [hunk 7]
```diff
@@ -584,7 +587,7 @@ IEnumerable<TResult> _(int? maxConcurrency, TaskScheduler scheduler, bool ordere
                         }
                     }
 
-                    if (holds?.Count > 0) // yield any withheld, which should be in order...
+                    if (holds != null && holds.Count > 0) // yield any withheld, which should be in order...
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply MoreLinq/Experimental/Await.cs]: Agree and this was rolled back with 8439ced2c9e0e79a5992e618c2e6b2e60c661a11.

File: MoreLinq/EndsWith.cs [hunk 8]
```diff
@@ -77,13 +77,13 @@ public static bool EndsWith<T>(this IEnumerable<T> first, IEnumerable<T> second,
             return second.TryGetCollectionCount() is {} secondCount
                    ? first.TryGetCollectionCount() is {} firstCount && secondCount > firstCount
                      ? false
-                     : Impl(second, secondCount)
-                   : Impl(secondList = second.ToList(), secondList.Count);
+                     : Impl(second, secondCount, comparer)
+                   : Impl(secondList = second.ToList(), secondList.Count, comparer);
 
-            bool Impl(IEnumerable<T> snd, int count)
+            bool Impl(IEnumerable<T> snd, int count, IEqualityComparer<T> cmp)
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply MoreLinq/EndsWith.cs]: Agree and this was rolled back as part of 3c6282e4447eb15c8029552befba80cdf6f24369.

File: MoreLinq/Collections/Dictionary.cs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            }
+        }
+
+        public bool TryGetValue(TKey key, [MaybeNullWhen(false)] out TValue value)
+        {
+            DefaultGuard();
+
+            if (key is null)
+            {
+                switch (_null)
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply MoreLinq/Collections/Dictionary.cs:62]: Not only do I find it more readable _and_ maintainable, it also eliminates using dammit (`!`) for the true-case. `_null` is an option disguised as an [_optuple_](https://github.com/atifaziz/Optuple) and the patterns make that more obvious. If we ever get [_partial active patterns_](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/active-patterns#partial-active-patterns) then it will become even more readable.

File: MoreLinq/Collections/Dictionary.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+    /// <summary>
+    /// A minimal <see cref="System.Collections.Generic.Dictionary{TKey,TValue}"/> wrapper that
+    /// allows null keys when <typeparamref name="TKey"/> is a
+    /// reference type.
+    /// </summary>
+
+    // Add members if and when needed to keep coverage.
+
+    struct Dictionary<TKey, TValue>
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply MoreLinq/Collections/Dictionary.cs:32]: I can agree with 3, but 1 and 2 have lost their argument with when value tuples were made mutable. See bullet point 2 of the following passage from the [remarks section](https://docs.microsoft.com/en-us/dotnet/api/system.valuetuple?view=netcore-3.1#remarks) of [`ValueTuple` documentation](https://docs.microsoft.com/en-us/dotnet/api/system.valuetuple?view=netcore-3.1):

> Value tuples are tuple types introduced in the .NET Framework 4.7 to provide the runtime implementation of [tuples in C#](https://docs.microsoft.com/en-us/dotnet/csharp/tuples) and struct tuples in F#. They differ from the tuple classes, such as [`Tuple<T1>`](https://docs.microsoft.com/en-us/dotnet/api/system.tuple-1?view=netcore-3.1), [`Tuple<T1,T2>`](https://docs.microsoft.com/en-us/dotnet/api/system.tuple-2?view=netcore-3.1), etc., as follows:
>
> - They are structures (value types) rather than classes (reference types).
> - They are mutable rather than read-only. That is, the value of tuple components can change.
> - Their data members, such as `Item1`, `Item2`, etc., are fields rather than properties.

I am not saying that this suddenly justifies making all `struct` mutable. Those have been around for a long time for sake of optimisation, such as collection enumerators. Here, this dictionary is a mere wrapper gateway to the underlying implementation, allowing nulls as an extension. It is also a very incomplete implementation as per the comment:

```c#
    // Add members if and when needed to keep coverage.
```

File: MoreLinq/CountBy.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (fsateler) [reply MoreLinq/CountBy.cs:93]: OK, it seems there is not much difference:

    public class Benchmarks
    {
        private readonly (bool, object)[] data;

        public Benchmarks() {
            data = new [] {
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
                (true, (object)"string"),
                (false, null),
            };
        }
        
        [Benchmark]
        public void MTest()
        {
            foreach (var d in data)
                M(d);
        }
        [Benchmark]
        public void NTest()
        {
            foreach (var d in data)
                N(d.Item1, d.Item2);
        }

        public void M((bool, object) x) {
            if (x is (true, {} obj)) {
                obj.GetHashCode();
            }
        }
        public void N(bool f, object obj) {
            if (f) {
                obj.GetHashCode();
            }
        }
    }


```
// * Summary *

BenchmarkDotNet=v0.11.5, OS=debian 
Intel Core i7-7600U CPU 2.80GHz (Kaby Lake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.1.302
  [Host]     : .NET Core 3.0.3 (CoreCLR 4.700.20.6603, CoreFX 4.700.20.6701), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.3 (CoreCLR 4.700.20.6603, CoreFX 4.700.20.6701), 64bit RyuJIT


| Method |     Mean |     Error |    StdDev |
|------- |---------:|----------:|----------:|
|  MTest | 134.4 ns | 0.4219 ns | 0.3947 ns |
|  NTest | 136.0 ns | 0.2768 ns | 0.2312 ns |
```

Let's merge as it is

File: MoreLinq/Collections/Dictionary.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (fsateler) [reply MoreLinq/Collections/Dictionary.cs:32]: I feel you did not answer the question "why a struct"? My impression is that things are usually classes by default, unless one needs either value semantics or the performance gains. Neither of these reasons is present here, so I wonder why did you go with struct.

File: MoreLinq/Collections/Dictionary.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (fsateler) [reply MoreLinq/Collections/Dictionary.cs:62]: OK. Perhaps it's just me that has not kept up with all the latest changes.

File: MoreLinq/Collections/Dictionary.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (atifaziz) [reply MoreLinq/Collections/Dictionary.cs:32]: The `struct`-based dictionary is not a new type or abstraction per se so normal considerations like classes-by-default or value semantics aren't being strictly applied. It's there solely to avoid an extra allocation so the answer is performance by way of avoiding allocation. If you think this is not a satisfactory justification than I am happy to go with a class instead.

[the remark this answers is not in the document]
Comment (fsateler) [reply MoreLinq/Collections/Dictionary.cs:32]: I think I do prefer that.

This is the only issue left. After this, LGTM

[the remark this answers is not in the document]
Comment (atifaziz) [reply MoreLinq/Collections/Dictionary.cs]: Done with 861585cbb9de86ab4136080698e34a25a186b941. If you could give it your final blessings by approving the PR then I'll go ahead and merge.

Thanks a ton for your review and feedback!
